psk_conn.go

```diff
@@ -37,14 +37,9 @@ func (c *pskConn) Read(out []byte) (int, error) {
 		c.readS20 = salsa20.New(c.psk, nonce)
 	}
 
-	maxn := uint32(len(out))
-	in := mpool.ByteSlicePool.Get(maxn).([]byte) // get buffer
-	defer mpool.ByteSlicePool.Put(maxn, in)      // put the buffer back
-
-	in = in[:maxn]            // truncate to required length
-	n, err := c.Conn.Read(in) // read to in
+	n, err := c.Conn.Read(out) // read to in
 	if n > 0 {
-		c.readS20.XORKeyStream(out[:n], in[:n]) // decrypt to out buffer
+		c.readS20.XORKeyStream(out[:n], out[:n]) // decrypt to out buffer
```

Comment:
This is the only real change I made. This *works*, I just have no idea why we didn't do this in the first place (I have a feeling it was a vague security concern but, at worse, we'll end up exposing encrypted bytes that have been sent over the network so I'm not exactly concerned).

Reply:
Because we're now reading the encrypted bytes onto the buffer that's passed in by the caller, right? How would that be insecure? With salsa20, if you know the encrypted bytes and the decrypted bytes, can you crack the key?

Reply:
No, you cannot.

Reply:
> With salsa20, if you know the encrypted bytes and the decrypted bytes, can you crack the key?

No. Really, this should be fine. According to @Kubuxu, the docs used to be unclear on whether or not using the same buffer was safe (from a logic standpoint).

Reply:
It was cleared up last year: https://github.com/golang/crypto/commit/74b34b9dd60829a9fcaf56a59e81c3877a8ecd2c#diff-f7d579dda2d6d9536050719abf92604aL31 in September, I was writing this in May.